atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepTaskRunner.java

```diff
@@ -176,200 +153,67 @@ private SweepResults doRun(TableReference tableRef,
+                .shouldCheckIfLatestValueIsEmpty(sweeper.shouldSweepLastCommitted())
+                .timestampsToIgnore(sweeper.getTimestampsToIgnore())
                 .build();
 
-        try (ClosableIterator<RowResult<Value>> valueResults = getValues(tableRef, range, sweepTs, sweeper);
-             ClosableIterator<RowResult<Set<Long>>> rowResults = keyValueService.getRangeOfTimestamps(
-                     tableRef, range, sweepTs)) {
-            EquivalenceCountingIterator<RowResult<Set<Long>>> rowResultTimestamps =
-                    new EquivalenceCountingIterator<>(rowResults, rowBatchSize, sameRowEquivalence());
-            PeekingIterator<RowResult<Value>> peekingValues = Iterators.peekingIterator(valueResults);
-
-            BatchingVisitable<CellAndTimestamps> cellsAndTimestamps = BatchingVisitableFromIterable
-                    .create(getTimestampsFromRowResultsIterator(() -> rowResultTimestamps));
-
-            final AtomicInteger totalCellsSwept = new AtomicInteger(0);
-            cellsAndTimestamps.batchAccept(
-                    cellBatchSize,
-                    thisBatch -> {
-                        CellsAndTimestamps thisBatchCells = CellsAndTimestamps.fromCellAndTimestampsList(thisBatch);
-                        int cellsSwept = sweepForCells(thisBatchCells,
-                                tableRef,
-                                sweeper,
-                                sweepTs,
-                                peekingValues,
-                                runType);
-                        totalCellsSwept.addAndGet(cellsSwept);
-                        return true;
-                    });
-
-            byte[] nextRow = rowResultTimestamps.size() < rowBatchSize ? null :
-                    RangeRequests.getNextStartRow(false, rowResultTimestamps.lastItem().getRowName());
+        SweepableCellFilter sweepableCellFilter = new SweepableCellFilter(transactionService, sweeper, sweepTs);
+        GetCandidateCellsForSweepingShim shim = new GetCandidateCellsForSweepingShim(keyValueService);
+        try (ClosableIterator<List<CandidateCellForSweeping>> candidates = shim.getCandidateCellsForSweeping(
+                    tableRef, request)) {
+            ExaminedCellLimit limit = new ExaminedCellLimit(startRow, batchConfig.maxCellTsPairsToExamine());
+            Iterator<BatchOfCellsToSweep> batchesToSweep = getBatchesToSweep(
+                        candidates, batchConfig, sweepableCellFilter, limit);
+            long totalCellTsPairsExamined = 0;
+            long totalCellTsPairsDeleted = 0;
+            byte[] lastRow = startRow;
+            while (batchesToSweep.hasNext()) {
+                BatchOfCellsToSweep batch = batchesToSweep.next();
+                if (runType == RunType.FULL) {
+                    totalCellTsPairsDeleted += sweepBatch(tableRef, batch.cells());
+                }
+                totalCellTsPairsExamined = batch.numCellTsPairsExaminedSoFar();
```

Comment:
Yes, we always return the cumulative number of examined cells from the beginning of the getCandidateCellsForSweeping() call